File: PokerGame/PotAlgo.cs
namespace PokerGame;

public static class PotAlgo
{
    public static bool EnableDebugLog { get; set; } = true;

    class ChipTracker(GamePlayer owner, int value, bool HasFolded)
    {
        public GamePlayer Owner { get; } = owner;
        public int Value { get; set; } = value;
        public bool HasFolded { get; } = HasFolded;

        public override string ToString()
        {
            return $"Owner: {Owner.Name} | Value: {Value} | Folded: {HasFolded}";
        }
    }

    public static List<Pot> GetPots(List<GamePlayer> players)
    {
        bool atLeastOneNonFolded = false;
        foreach (var item in players)
        {
            if (!item.HasFolded)
            {
                atLeastOneNonFolded = true;
                break;
            }
        }
        if (atLeastOneNonFolded == false) throw new InternalPokerEngineException("GetPots() called when all players have folded.");

        List<ChipTracker> trackers = [];
        foreach (GamePlayer p in players)
        {
            if (p.Bet != 0)
            {
                if(p.Bet < 0) throw new InternalPokerEngineException($"Player {p.Name} has a negative bet value.");
                trackers.Add(new ChipTracker(p, p.Bet, p.HasFolded));
            }
        }

        if (trackers.Count == 0) throw new InternalPokerEngineException("GetPots() called when no players have bet anything.");

        if (EnableDebugLog)
        {
            Console.WriteLine("Trackers:");
            foreach (ChipTracker t in trackers)
            {
                Console.WriteLine(t);
            }
            Console.WriteLine();
        }

        return SplitPot(trackers);
    }

    private static List<Pot> SplitPot(List<ChipTracker> trackers)
    {
        // end condition
        if (trackers.Count == 0) {
            if (EnableDebugLog) Console.WriteLine("End of Recursion.");
            return [];
        }

        // pot splitting logic
        Pot pot;
        int min = GetMin(trackers);
        int potTotal = 0;
        int foldedTotal = 0;
        List<GamePlayer> potPlayers = [];

        // loop through trackers and decrease value
        foreach (ChipTracker t in trackers)
        {
            if (t.HasFolded)
            {
                if (t.Value <= min)
                {
                    foldedTotal += t.Value;
                    t.Value = 0;
                }
                else
                {
                    foldedTotal += min;
                    t.Value -= min;
                }
            }
            else
            {
                potPlayers.Add(t.Owner);
                potTotal += min;
                t.Value -= min;
            }
        }

        pot = new Pot(potTotal + foldedTotal, potPlayers);
        if (EnableDebugLog)
        {
            Console.WriteLine($"Current Number of Trackers: {trackers.Count}");
            Console.WriteLine("Pot in Recursion:");
            Console.WriteLine(pot);
            Console.WriteLine();
        }

        // prepare trackers for next recursion
        trackers.RemoveAll(t => t.Value == 0);

        // we combine all the pots
        List<Pot> pots = [pot];
        pots.AddRange(SplitPot(trackers));
        return pots;
    }

    private static int GetMin(List<ChipTracker> trackers)
    {
        int min = int.MaxValue;
        bool found = false;

        foreach (var t in trackers)
        {
            if (t.HasFolded) continue;

            found = true;
            if (t.Value < min) min = t.Value;
        }

        if (!found) throw new InternalPokerEngineException("GetMin() called with no non-folded trackers remaining.");

        return min;
    }

}

File: pot_algo.go
package pokergame

import "fmt"

var EnablePotDebugLog = true

type chipTracker struct {
	Owner     *GamePlayer
	Value     int
	HasFolded bool
}

func (t *chipTracker) String() string {
	return fmt.Sprintf("Owner: %s | Value: %d | Folded: %t", t.Owner.Name, t.Value, t.HasFolded)
}

func GetPots(players []*GamePlayer) ([]*Pot, error) {
	atLeastOneNonFolded := false
	for _, item := range players {
		if !item.HasFolded {
			atLeastOneNonFolded = true
			break
		}
	}
	if atLeastOneNonFolded == false {
		return nil, newInternalPokerGameError("GetPots() called when all players have folded.")
	}

	trackers := []*chipTracker{}
	for _, p := range players {
		if p.Bet != 0 {
			if p.Bet < 0 {
				return nil, newInternalPokerGameError("Player %s has a negative bet value.", p.Name)
			}
			trackers = append(trackers, &chipTracker{Owner: p, Value: p.Bet, HasFolded: p.HasFolded})
		}
	}

	if len(trackers) == 0 {
		return nil, newInternalPokerGameError("GetPots() called when no players have bet anything.")
	}

	if EnablePotDebugLog {
		fmt.Println("Initial Chip Trackers:")
		for _, t := range trackers {
			fmt.Println(t)
		}
		fmt.Println()
	}

	return splitPot(trackers)
}

func splitPot(trackers []*chipTracker) ([]*Pot, error) {
	// end condition
	if len(trackers) == 0 {
		if EnablePotDebugLog {
			fmt.Println("End of Recursion.")
		}
		return []*Pot{}, nil
	}

	// pot splitting logic
	min, err := getMinBet(trackers)
	if err != nil {
		return nil, err
	}
	potTotal := 0
	foldedTotal := 0
	potPlayers := []*GamePlayer{}

	// loop through trackers and remove value
	for _, t := range trackers {
		if t.HasFolded {
			if t.Value <= min {
				foldedTotal += t.Value
				t.Value = 0
			} else {
				foldedTotal += min
				t.Value -= min
			}
		} else {
			potPlayers = append(potPlayers, t.Owner)
			potTotal += min
			t.Value -= min
		}
	}

	pot := NewPot(potTotal+foldedTotal, potPlayers)
	if EnablePotDebugLog {
		fmt.Println("Pot in Recursion:")
		fmt.Println(pot)
		fmt.Printf("Current Number of Trackers: %d\n", len(trackers))
		fmt.Println()
	}

	// prepare trackers for next recursion
	nextTrackers := trackers[:0]
	for _, t := range trackers {
		if t.Value != 0 {
			nextTrackers = append(nextTrackers, t)
		}
	}

	// we combine all the pots
	pots := []*Pot{pot}
	nextPots, err := splitPot(nextTrackers)
	if err != nil {
		return nil, err
	}
	pots = append(pots, nextPots...)
	return pots, nil
}

func getMinBet(trackers []*chipTracker) (int, error) {
	min := int(^uint(0) >> 1)
	found := false

	for _, t := range trackers {
		if t.HasFolded {
			continue
		}

		found = true
		if t.Value < min {
			min = t.Value
		}
	}

	if !found {
		return 0, newInternalPokerGameError("GetMin() called with no non-folded trackers remaining. There should be at least one non-folded player, with the highest bet.")
	}

	return min, nil
}
